### PR TITLE
MemoryManager class

### PR DESCRIPTION
**Pull request by RubenNatanael**
The class implements methods to:

- Read from a file
- Read from a zone of memory
- Write in to a file
- Write in an address of memory
- Verification of address and memory availability

The class architecture is:

1. Singleton PD. The object is initialize with the address in the requestDownload service. In order to be accessible in the transfer data service, using the singleton PD the object returned is the same.
2. Static methods for readFromAddress(), writeToFile() and readBinary(), is not necessary to create an object to make this actions.

### CODE REVIEW

**Review by DanielConstantinescu — APPROVED** [review]

File: src/utils/include/MemoryManager.h [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+         * @return int 
+         */
+        int to_int(std::string number);
+
+        /**
+         * @brief Method to check if the address is available
+         * 
+         * @param address 
+         * @return true 
+         * @return false
```
Comment (Miruna-Maria) [inline src/utils/include/MemoryManager.h:89]: please choose true or false

File: src/utils/include/MemoryManager.h [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+         * @return false 
+         */
+        bool availableAddress(off_t address);
+
+        /**
+         * @brief Method to check if the amount of memory is available
+         * 
+         * @param size_of_data 
+         * @return true 
+         * @return false
```
Comment (Miruna-Maria) [inline src/utils/include/MemoryManager.h:98]: same here

File: src/utils/src/MemoryManager.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    return this->path;
+}
+
+std::string MemoryManager::runCommand(char command[])
+{
+    char buffer[128];
+    std::string result = "";
+    FILE* pipe = popen(command, "r");
+    if (!pipe) throw std::runtime_error("popen() failed!");
+    try {
```
Comment (Miruna-Maria) [inline src/utils/src/MemoryManager.cpp:67]: please start with { on next row

File: src/utils/src/MemoryManager.cpp [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        while (fgets(buffer, sizeof buffer, pipe) != NULL) {
```
Comment (Miruna-Maria) [inline src/utils/src/MemoryManager.cpp:68]: same here

File: src/utils/src/MemoryManager.cpp [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+            result += buffer;
+        }
+    } catch (...) {
```
Comment (Miruna-Maria) [inline src/utils/src/MemoryManager.cpp:71]: here for catch and also for {

File: src/utils/src/MemoryManager.cpp [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    return uint8Buffer;
+}
+
+void print_bin_data(std::vector<uint8_t> &data)
+{
+    int i=0;
+    for(auto element : data)
+    {
+        std::cout<<std::hex<< std::setw(2) << std::setfill('0') <<(unsigned int)element<< " ";
+        if ((i + 1) % 16 == 0) {
```
Comment (Miruna-Maria) [inline src/utils/src/MemoryManager.cpp:258]: here also

**Review by Miruna-Maria — APPROVED** [review]

**Pull request merged**